### PR TITLE
add Intel mpi package and add MPI wrappers to Intel parallel studio

### PR DESCRIPTION
**Pull request by davydden**
as per discussion in https://groups.google.com/forum/#!topic/spack/SDR04CNRoyQ and https://groups.google.com/forum/#!topic/spack/8ep-wfq9AVw

testing 
- [x] `spack install hdf5%gcc+mpi^intel-mpi`
- [x] `spack install hdf5%intel+mpi^intel-mpi`

on CentOS 7 cluster with  (setup with `spack compiler find` after `module load intelmpi`)
```
compilers:
- compiler:
    environment: {}
    extra_rpaths: []
    flags: {}
    modules: []
    operating_system: centos7
    paths:
      cc: /usr/bin/gcc
      cxx: /usr/bin/g++
      f77: /usr/bin/gfortran
      fc: /usr/bin/gfortran
    spec: gcc@4.8.5
- compiler:
    environment: {}
    extra_rpaths: []
    flags: {}
    modules: []
    operating_system: centos7
    paths:
      cc: /apps/intel/ComposerXE2016/compilers_and_libraries_2016.3.210/linux/bin/intel64/icc
      cxx: /apps/intel/ComposerXE2016/compilers_and_libraries_2016.3.210/linux/bin/intel64/icpc
      f77: /apps/intel/ComposerXE2016/compilers_and_libraries_2016.3.210/linux/bin/intel64/ifort
      fc: /apps/intel/ComposerXE2016/compilers_and_libraries_2016.3.210/linux/bin/intel64/ifort
    spec: intel@16.0.3
    target: x86_64
```
and
```
  intel-mpi:
    version: ['5.1.3']
    paths:
      intel-mpi@5.1.3%intel@16.0.3: /apps/intel/mpi/5.1.3.210/
      intel-mpi@5.1.3%gcc@4.8.5: /apps/intel/mpi/5.1.3.210/
    buildable: False
```

### CODE REVIEW

**Comment by davydden** [comment]
can't test the compilation with `%intel` due to 
```
Compiler error reporting is too harsh for ./configure (perhaps remove -Werror).
** ./configure aborting.
```
in `zlib`.  If anyone has any ideas -- let me know...

**Comment by davydden** [comment]
as for the flake8 issues, it appears that i can't remove those `import` as they are needed for the parent package.

File: var/spack/repos/builtin/packages/intel-mpi/package.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    version('2017.1', '8d30a63674fe05f17b0a908a9f7d54403018bfed2de03c208380b171ab99be82',
+            url='http://registrationcenter-download.intel.com/akdlm/irc_nas/tec/11014/l_mpi_2017.1.132.tgz')
+    version('5.1.3', '544f4173b09609beba711fa3ba35567397ff3b8390e4f870a3307f819117dd9b',
+            url='http://registrationcenter-download.intel.com/akdlm/irc_nas/tec/9278/l_mpi_p_5.1.3.223.tgz')
+
+    provides('mpi')
+
+    def install(self, spec, prefix):
+        # FIXME:
+        raise RuntimeError('Install method is not implemented yet')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/intel-mpi/package.py]: Can you implement this? It shouldn't be that different from the other Intel packages.

Comment (davydden) [inline var/spack/repos/builtin/packages/intel-mpi/package.py]: i added that, but i can't test it as I don't have a licence.

File: var/spack/repos/builtin/packages/intel-mpi/package.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    version('5.1.3', '544f4173b09609beba711fa3ba35567397ff3b8390e4f870a3307f819117dd9b',
+            url='http://registrationcenter-download.intel.com/akdlm/irc_nas/tec/9278/l_mpi_p_5.1.3.223.tgz')
+
+    provides('mpi')
+
+    def install(self, spec, prefix):
+        self.intel_prefix = prefix
+        IntelInstaller.install(self, spec, prefix)
+
+
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/intel-mpi/package.py]: Too many blank lines

File: var/spack/repos/builtin/packages/intel-mpi/package.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+        spack_env.set('I_MPI_CC', spack_cc)
+        spack_env.set('I_MPI_CXX', spack_cxx)
+        spack_env.set('I_MPI_F77', spack_fc)
+        spack_env.set('I_MPI_F90', spack_f77)
+
+    def setup_dependent_package(self, module, dep_spec):
+        # Check for presence of bin64 or bin directory
+        if os.path.isdir(self.prefix.bin):
+            bindir = self.prefix.bin
+        elif os.path.isdir(join_path(self.prefix, 'bin64')):
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/intel-mpi/package.py]: You might as well add `prefix.bin64` to https://github.com/LLNL/spack/blob/develop/lib/spack/spack/util/prefix.py#L63

File: var/spack/repos/builtin/packages/intel-mpi/package.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+            bindir = join_path(self.prefix, 'bin64')
+        else:
+            raise "No suitable bindir found"
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/intel-mpi/package.py]: Don't you need an error class?

Comment (davydden) [inline var/spack/repos/builtin/packages/intel-mpi/package.py]: you are probably right, i blindly copy-pasted this from https://github.com/LLNL/spack/pull/2980

**Comment by lee218llnl** [comment]
@davydden, looks like you have some version lines that are too long in the intel-mpi/package.py file.

**Comment by adamjstewart** [comment]
It's actually the hash causing the version line to be too long, not the URL. We only exempt `version(...)` lines, not `version(...,` lines. We could change that.

**Comment by davydden** [comment]
i updated, current flake8 issues are not related

**Comment by alalazo** [comment]
@davydden I restarted the `flake8` test on Travis and it succeeded.

**Pull request merged**